3.10 Release notes (#6588)
medic/cht-core: release-notes/docs/3.10.0.md

# 3.10.0 Release Notes

- [Known issues](#known-issues)
- [Upgrade notes](#upgrade-notes)
- [Supporting remote first-time login](#supporting-remote-first-time-login)
- [Supporting linked docs](#supporting-linked-docs)
- [Privacy Policies](#privacy-policies)
- [More granular replication depth configuration for offline users](#more-granular-replication-depth-configuration-for-offline-users)
- [Improved geolocation data saved for reports](#improved-geolocation-data-saved-for-reports)
- [Customizable header tabs icons](#customizable-header-tabs-icons)
- [Improved performance of processing deleted docs](#improved-performance-of-processing-deleted-docs)
- [And more...](#and-more)

## Known issues

None.

## Upgrade notes

> Sentinel no longer advances its transition processing seq in `_local/sentinel-meta-data`. Instead, it advances the transition processing seq in `_local/transitions-seq`. Additionally, a new document `_local/background-seq` advances the deletion processing seq. 

Reviewer: I'm not sure this should be mentioned, or at least, not mentioned here. It sounds like we require people to do something before upgrade but that's not the case. Maybe it's a breaking change?

Author: It kindof is a breaking change. Often, TLs and SREs need to advance Sentinel process seq, so I think we should mention this somewhere.

Reviewer: Fair enough. Let's put it in the breaking changes section.

### Breaking changes

#### Updates to meta databases replication

The replication of users meta databases to the conglomerate, `medic-users-meta` database, is no longer configurable. 
This task now runs every day, at 2am UTC, and replicates `feedback` and `telemetry` documents to `medic-users-meta` database.

#### Android wrapper update required for new features

> Some features in 3.10 are only fully functional while using the latest release of medic-android.

Reviewer: We should specify which version (when it's released) and ideally specify which features require it.

### Supported software

There are no required changes to the [supported software matrix](https://github.com/medic/medic-docs/blob/master/installation/supported-software.md) from `3.0.0`.

| Node | CouchDB | Browsers | SMS bridge | Android | medic-android | medic-couch2pg |
|----|----|----|----|----|----|---|
| 8.11+ | 2.1+ | Chrome 53+, Firefox latest | medic-gateway | 4.4+ | 0.4.5+ ???? | 3.0+ |

> // todo medic-android version?? 

Reviewer: It's backwards compatible, right? So I don't think we need to bump the minimum medic-android version. We just need to make it clear that certain features won't work without it.

#### Supporting remote first-time login

3.10.0 adds a new authentication mechanism that doesn't require users to know their username or password. Users can now access the app for the first time by clicking on a token authentication link that they receive by SMS. Admins can enable or disable token authentication for any user. This feature requires the latest version of the `medic-android` app.  
See [the documentation](https://docs.communityhealthtoolkit.org/apps/reference/api/#login-by-sms) for more information.    

> // todo picture?

Reviewer: Yes please!

It might also be worth saying something about how the android app now supports weblinks in general?

Author: I've added both a picture and note about links.

// todo picture?

#### More granular replication depth configuration for offline users

This release adds a new `replication_depth` configuration option that limits the depth of replicated reports. This allows app builders to configure roles that replicate deeper contacts, can report against these contacts, but will not download reports about these contacts that have been created by other users.  
See [the documentation](https://docs.communityhealthtoolkit.org/apps/guides/performance/replication/#report-depth) for more information.

#### Improved geolocation data

3.10.0 improves the way geolocation information is obtained, from requesting an instant position to allowing up to 30 seconds to get a position lock. Additionally, the `geolocation` property is now updated every time a report is edited, and the all previous geolocation entries are saved in a log within the same report.  
This change also supports the latest updates in `medic-android`, where Location permission is requested at runtime, when editing/creating a report, rather than demanded when starting the app for the first time.   

> #### Customizable header tabs icons

Reviewer: Let's not include this one...

As of 3.10.0, header tabs icons can be customized, allowing any [FontAwesome](https://fontawesome.com/v4.7.0/) icon or [resource icon](https://docs.communityhealthtoolkit.org/design/icons/) in SVG format to be used as a tab icon.
Tabs icons can be configured using `medic-conf` or from a new App Management page.  
See [the documentation](https://docs.communityhealthtoolkit.org/apps/reference/app-settings/header_tabs/) for more information.  
// todo picture? 

#### Improved performance of processing deleted docs

Because processing document deletes, specifically deleting read docs, is relatively time-consuming, increasing linearly by the number of users that exist on the instance, deletion processing is now split into: 
- generating tombstones: no changes, tombstones are generated like before, within the Sentinel transition processing feed.
- deleting read docs and info docs: done within a new Sentinel scheduled task that doesn't block the main transitions processing feed

> Additionally, Sentinel scheduled tasks are no longer executed sequentially, they are now executed in parallel, keeping track of which tasks are running. Every 5 minutes, every scheduled task that is not already running is started.

Reviewer: Is there any way to show how much quicker this is?

Author: I don't think anybody did any kind of benchmarking. Would you like me to benchmark ? (I can use the exact numbers used in the issue description: 5000 deletes for 1000 users).

Reviewer: If it's not too much work that'd be perfect.

Author: Oh man ... I lowered the number to 500 after estimating it would take 8 hours. I did the test with 5000 docs in the new version, tho.

- [cht-core#6313](https://github.com/medic/cht-core/issues/6313): Use the browser language as the default language
- [cht-core#6370](https://github.com/medic/cht-core/issues/6370): Replication "isSensitive" logic prevents "boss" from creating any report for users directly under them in hierarchy
- [cht-core#6386](https://github.com/medic/cht-core/issues/6386): Error on report not cleared
- [cht-core#6404](https://github.com/medic/cht-core/issues/6404): User name autofilled with "medic" 
- [cht-core#6433](https://github.com/medic/cht-core/issues/6433): Infinite spinners when loading tabs as admin
- [cht-core#6502](https://github.com/medic/cht-core/issues/6502): Guided tour erratic when user has no access to any tab
- [cht-core#6519](https://github.com/medic/cht-core/issues/6519): Upgrading doesn't apply replication configuration which means `users-meta` is empty
- [cht-core#6562](https://github.com/medic/cht-core/issues/6562): Muso production users have multiple task documents in non-terminal state with the same emission id
- [cht-core#6567](https://github.com/medic/cht-core/issues/6567): Security issue medic-infrastructure#147

> - [cht-core#6568](https://github.com/medic/cht-core/issues/6568): Security issue medic-infrastructure#190

Reviewer: I think we should list security improvements separately... maybe we need a new Type label?

Anyway let's disclose these a little better - right now nobody outside of the org can see what we did. Think about what we can say without disclosing too much about the actual vulnerability?

- [cht-core#6583](https://github.com/medic/cht-core/issues/6583): Users are forced to login one year after they last logged in
- [horticulturalist#57](https://github.com/medic/horticulturalist/issues/57): Horti doesn't wam all views when deploying 3.9.x+
- [medic-conf#290](https://github.com/medic/medic-conf/issues/290): Tests fail locally when Bash completion is not setup
- [medic-conf#315](https://github.com/medic/medic-conf/issues/315): Error running csv-to-docs with null properties
- [medic-conf#322](https://github.com/medic/medic-conf/issues/322): Missing optional appliesIf attribute on contact summary cards causes crash
- [medic-conf#332](https://github.com/medic/medic-conf/issues/332): Uploading of multimedia is extremely limited in medic-conf

### Technical issues

- [cht-core#6369](https://github.com/medic/cht-core/issues/6369): Detect unhandled rejections in unit tests
- [cht-core#6379](https://github.com/medic/cht-core/issues/6379): Figure out how to deploy RapidPro in production
- [cht-core#6439](https://github.com/medic/cht-core/issues/6439): Auth service unit test is flaky
- [cht-core#6463](https://github.com/medic/cht-core/issues/6463): Bump dependencies for 3.10

> - [medic-android#111](https://github.com/medic/medic-android/issues/111): Compliance with Play Store developer policies for PII collection disclosure

Reviewer: This shouldn't be a technical issue...

Author: Uhm, ok... improvement then maybe? (you tagged it technical issue)

Author: The more I think about it, the more I think the technical  issue label is accurate. 

Existing users (that have already installed the app) should notice no difference (because they've already granted permissions when first running or upon installation?).

Reviewer: > you tagged it technical issue

Hah. Well one of me is wrong. I'm not too bothered which. It's sort of an App Builder's feature? Anyway, just ignore me.

# 3.10.0 Release Notes

- [Known issues](#known-issues)
- [Upgrade notes](#upgrade-notes)
- [Breaking changes](#breaking-changes)
- [Supporting remote first-time login](#supporting-remote-first-time-login)
- [Supporting linked docs](#supporting-linked-docs)
- [Privacy Policies](#privacy-policies)
- [More granular replication depth configuration for offline users](#more-granular-replication-depth-configuration-for-offline-users)
- [Improved geolocation data saved for reports](#improved-geolocation-data-saved-for-reports)
- [Improved performance of processing deleted docs](#improved-performance-of-processing-deleted-docs)
- [And more...](#and-more)

## Known issues

None.

## Upgrade notes 

### Breaking changes

#### Updates to meta databases replication

The replication of users meta databases to the conglomerate, `medic-users-meta` database, is no longer configurable. 
This task now runs every day, at 2am UTC, and replicates `feedback` and `telemetry` documents to `medic-users-meta` database.

#### Android wrapper update required for new features

> 

Reviewer 2: Should we mention GPS improvement as well since it requires a new wrapper at time of upgrade to be functional?

Author: The cht-core side GPS is compatible with previous versions of the android wrapper.

This feature can have a multitude of applications, from providing a way to access an arbitrary contact's phone number when resolving an SMS recipient to linking a delivery report to a child person document (`linked_docs` can also be used in `contact-summary`).   
//todo documentation link

#### Privacy Policies

App builders can now configure privacy policies for every language. When a privacy policy is configured for a language, users who load the app in this language are prompted to accept the policy, and are blocked from using the app until they have accepted. Privacy policies can be configured using `medic-conf` or from a new App Management page.    
See [the documentation](https://docs.communityhealthtoolkit.org/apps/guides/security/privacy-policy/) for more information.  
![privacy policies](./images/3.10-privacy-policies.png)

#### More granular replication depth configuration for offline users

> This release adds a new `replication_depth` configuration option that limits the depth of replicated reports. This allows app builders to configure roles that replicate deeper contacts, can report against these contacts, but will not download reports about these contacts that have been created by other users.  

Reviewer 2: ```suggestion
This release adds a new `replication_depth` configuration option that limits the depth of replicated reports. This allows app builders to configure roles that can replicate deeper contacts and report against these contacts. This configuration option does not allow for downloading reports about these contacts that have been created by other users.  
```

This change also supports the latest updates in `medic-android`, where Location permission is requested at runtime, when editing/creating a report, rather than demanded when starting the app for the first time.   

#### Improved performance of processing deleted docs

Because processing document deletes, specifically deleting read docs, is relatively time-consuming, increasing linearly by the number of users that exist on the instance, deletion processing is now split into: 
- generating tombstones: no changes, tombstones are generated like before, within the Sentinel transition processing feed.
- deleting read docs and info docs: done within a new Sentinel scheduled task that doesn't block the main transitions processing feed

| Version | Number of users | Number of deleted docs | Time to process |
| ------- | --------------- | ---------------------- | --------------- |
| 3.9.0 | 1000 | 500 | 60 minutes | 
| 3.10.0 | 1000 | 500 | 45 seconds |  

> | 3.10.0 | 1000 | 5000 | 11 minutes |  

Reviewer: At first glance I thought we'd regressed because 11 minutes > 45 seconds... I think we should drop the third line so we're comparing apples to apples.

Also because the unit changes it looks like a minor improvement (eg: 60 somethings to 45 somethings), so make the units all seconds IMO.

Finally, because it breaks up the big wall of text, and because the improvement is so significant, I think we should convert this to a graph so it's visually obvious that 3.10.0 is 80 times quicker int his use case.

Author: > 60 somethings to 45 somethings
lol reading be hard 
I'll make a graph :)

# 3.10.0 Release Notes

- [Known issues](#known-issues)
- [Upgrade notes](#upgrade-notes)
- [Breaking changes](#breaking-changes)
- [Supported software](#supported-software)
- [Supporting remote first-time login](#supporting-remote-first-time-login)
- [Supporting linked docs](#supporting-linked-docs)
- [Privacy Policies](#privacy-policies)
- [More granular replication depth configuration for offline users](#more-granular-replication-depth-configuration-for-offline-users)
- [Improved geolocation data](#improved-geolocation-data)
- [Improved performance of processing deleted docs](#improved-performance-of-processing-deleted-docs)
- [And more...](#and-more)

## Known issues

None.

## Upgrade notes 

### Breaking changes

#### Updates to meta databases replication

The replication of users meta databases to the conglomerate, `medic-users-meta` database, is no longer configurable. 
This task now runs every day, at 2am UTC, and replicates `feedback` and `telemetry` documents to `medic-users-meta` database.

#### Android wrapper update required for new features

Supporting remote first-time login is only fully functional while using the latest release of medic-android.

> // todo version number

Author: I haven't released the latest version of medic-android yet.
